Sources/ComposableArchitectureMacros/ReducerMacro.swift

```diff
@@ -0,0 +1,172 @@
+import SwiftSyntaxMacroExpansion
+import SwiftSyntaxMacros
+
+public enum ReducerMacro {
+}
+
+extension ReducerMacro: ExtensionMacro {
+  public static func expansion<D: DeclGroupSyntax, T: TypeSyntaxProtocol, C: MacroExpansionContext>(
+    of node: AttributeSyntax,
+    attachedTo declaration: D,
+    providingExtensionsOf type: T,
+    conformingTo protocols: [TypeSyntax],
+    in context: C
+  ) throws -> [ExtensionDeclSyntax] {
+    if let inheritanceClause = declaration.inheritanceClause,
+      inheritanceClause.inheritedTypes.contains(
+        where: {
+          ["Reducer"].withQualified.contains($0.type.trimmedDescription)
+        }
+      )
+    {
+      return []
+    }
+    let ext: DeclSyntax =
+      """
+      extension \(type.trimmed): ComposableArchitecture.Reducer {}
+      """
+    return [ext.cast(ExtensionDeclSyntax.self)]
+  }
+}
+
+extension ReducerMacro: MemberAttributeMacro {
+  public static func expansion<D: DeclGroupSyntax, M: DeclSyntaxProtocol, C: MacroExpansionContext>(
+    of node: AttributeSyntax,
+    attachedTo declaration: D,
+    providingAttributesFor member: M,
+    in context: C
+  ) throws -> [AttributeSyntax] {
+    if let enumDecl = member.as(EnumDeclSyntax.self) {
+      var attributes: [String] = []
+      switch enumDecl.name.text {
+      case "State":
+        attributes = ["CasePathable", "dynamicMemberLookup"]
+      case "Action":
+        attributes = ["CasePathable"]
+      default:
+        break
+      }
+      for attribute in enumDecl.attributes {
+        guard
+          case let .attribute(attribute) = attribute,
+          let attributeName = attribute.attributeName.as(IdentifierTypeSyntax.self)?.name.text
+        else { continue }
+        attributes.removeAll(where: { $0 == attributeName })
+      }
+      return attributes.map {
+        AttributeSyntax(attributeName: IdentifierTypeSyntax(name: .identifier($0)))
+      }
+    } else if let property = member.as(VariableDeclSyntax.self),
+      property.bindingSpecifier.text == "var",
+      property.bindings.count == 1,
+      let binding = property.bindings.first,
+      let identifier = binding.pattern.as(IdentifierPatternSyntax.self)?.identifier,
+      identifier.text == "body",
+      case .getter = binding.accessorBlock?.accessors
+    {
+      if let reduce = declaration.memberBlock.members.first(where: {
+        guard
+          let method = $0.decl.as(FunctionDeclSyntax.self),
+          method.name.text == "reduce",
+          method.signature.parameterClause.parameters.count == 2,
+          let state = method.signature.parameterClause.parameters.first,
+          state.firstName.text == "into",
+          state.type.as(AttributedTypeSyntax.self)?.specifier?.text == "inout",
+          method.signature.parameterClause.parameters.last?.firstName.text == "action",
+          method.signature.effectSpecifiers == nil,
+          method.signature.returnClause?.type.as(IdentifierTypeSyntax.self) != nil
```

Comment:
While the current might be more complex, I'm comfortable with the checks, and I think it's checking just what needs to be checked and nothing more. Folks can technically define a `reduce` that matches the protocol conformance but uses a custom type for `State` or `Action` (since those are just associated types), and one could also define custom secondary names. An extreme example:

```swift
func reduce(
  into myState: MyState,
  action myAction: MyAction<Something>
) -> EffectOf<MyFeature>
```
```

Reply:
> In this scenario, wouldn't the private `reduce` method, be detected instead of the one intended from the protocol?

That's not valid Swift, so the compiler would complain at the conformance level:

> 🛑 Method 'reduce(into:action:)' must be as accessible as its enclosing type because it matches a requirement in protocol 'Reducer'

Reply:
> > In this scenario, wouldn't the private `reduce` method, be detected instead of the one intended from the protocol?
> 
> That's not valid Swift, so the compiler would complain at the conformance level:
> 
> > 🛑 Method 'reduce(into:action:)' must be as accessible as its enclosing type because it matches a requirement in protocol 'Reducer'

OK, I think I understand what happened here. In theory, the following code should indeed result in an error, but... it seems that, thanks to the implementation of the `Reduce` protocol being handled by a macro, the code works just fine 🤯

```swift
@Reducer
struct MyTest {
  struct State {}

  enum Action {
    case none
  }

  var body: some Reducer<State, Action> { Reduce { _, _ in .none } }

  private func reduce(into state: inout State, action: Action) -> Effect<Action> {
    .none
  }
```

<img width="618" alt="Screenshot 2023-11-11 at 9 12 05 PM" src="https://github.com/pointfreeco/swift-composable-architecture/assets/44930823/b4d47676-71f9-4b86-a8e2-2011846f2404">

Reply:
> While the current might be more complex, I'm comfortable with the checks, and I think it's checking just what needs to be checked and nothing more. Folks can technically define a `reduce` that matches the protocol conformance but uses a custom type for `State` or `Action` (since those are just associated types), and one could also define custom secondary names. An extreme example:
> 
> ```swift
> func reduce(
>   into myState: MyState,
>   action myAction: MyAction<Something>
> ) -> EffectOf<MyFeature>
> ```

Sorry, I'm a bit confused now. So, is your intention that every method like this should produce an error, right? Even ones with different types, like this example:

```swift
@Reducer
struct MyTest {
  struct State {}

  enum Action {
    case none
  }

  enum LocalAction {
    case some
  }

  var body: some Reducer<State, Action> { Reduce { _, _ in .none } }

  private func reduce(into state: inout State, action: LocalAction) -> Effect<Action> {
    .none
  }
}
```
Is that correct?

Reply:
@Matejkob I think there's only so much we can do from the macro to detect that a method matches the conformance. Someone could type alias the `Effect` return type so we can't do any checking there, and while we could traverse the reducer to see if there are `State` or `Action` types/aliases and use that to silence the warning for your above example, I'm not sure it's worth the extra work when at the end of the day, defining a method named `reduce` is confusing enough that we think a rename is warranted for folks to silence the warning. In fact, I did just add a syntax visitor that tries to fix-it a rename.

Reply:
@stephencelis Thanks for your explanation. I understand the challenges with the macro approach and agree that a rename is a more straightforward solution. Appreciate your work on this and the helpful response!